### PR TITLE
feat: create user profile on sign up

### PR DESCRIPTION
**Pull request by TheCodeGhinux**
# What does this PR do?
This PR implements creating user profile:
* Added POST `/api/v1/auth/register`
* Error Handling: Returns specific HTTP status codes and messages based on various scenarios.

 ### Endpoint: Creates a user acount and automatically creates a user profile

# Checklist:
- [x] Implemented POST `/api/v1/auth/register`  to retrieve the current user's profile settings.
- [x] Implemented an endpoint to create user profile.
- [x] Return 201 CREATED with the user's profile.
- [x] Return 400 Bad Request if the process is unsuccessful.
- [x] Tested endpoint functionality to ensure it correctly handles various scenarios.
- [x] Added appropriate HTTP status codes and response messages for edge cases.
- [x] Added unit tests for the `Profile Service`.

### CODE REVIEW

**Review by Heba-WebDev — APPROVED** [review]

File: package.json [hunk 1]
```diff
@@ -93,7 +93,7 @@
     "ts-node": "^10.9.2",
     "ts-node-dev": "^2.0.0",
     "tsconfig-paths": "^4.2.0",
-    "typescript": "^5.5.4"
+    "typescript": "^5.5.3"
```
Comment (Cobidex) [inline package.json:96]: why down grading the typescript version?

File: src/modules/auth/auth.service.ts [hunk 2]
```diff
@@ -77,6 +80,13 @@ export default class AuthenticationService {
         };
       }
 
+      // Create a profile with the user's email
```
Comment (Cobidex) [inline src/modules/auth/auth.service.ts:83]: remove dangling comment

File: src/modules/profile/dto/create-profile.dto.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { IsEmail, IsOptional, IsString } from 'class-validator';
+import { User } from '../../user/entities/user.entity';
+
+export class CreateProfileDto {
+  @IsOptional()
+  @IsString()
+  username?: string;
```
Comment (Cobidex) [inline src/modules/profile/dto/create-profile.dto.ts:7]: input dtos should be readonly by default

File: src/modules/profile/profile.service.ts [hunk 4]
```diff
@@ -11,6 +12,36 @@ export class ProfileService {
...
 
+  async createProfile(createProfileDto: CreateProfileDto, userId: string) {
+    try {
+      // Fetch the user entity by userId
+      const user = await this.userRepository.findOne({ where: { id: userId } });
+      if (!user) {
+        throw new BadRequestException('User not found');
+      }
+
+      // Check if a profile for this user already exists
```
Comment (Cobidex) [inline src/modules/profile/profile.service.ts:23]: remove dangling comments

File: src/modules/profile/profile.service.ts [hunk 5]
```diff
@@ -11,6 +12,36 @@ export class ProfileService {
     @InjectRepository(User) private userRepository: Repository<User>
   ) {}
 
+  async createProfile(createProfileDto: CreateProfileDto, userId: string) {
+    try {
+      // Fetch the user entity by userId
```
Comment (Cobidex) [inline src/modules/profile/profile.service.ts:17]: dangling comment

File: src/modules/profile/profile.service.ts [hunk 6]
```diff
@@ -11,6 +12,36 @@ export class ProfileService {
...
+      const user = await this.userRepository.findOne({ where: { id: userId } });
+      if (!user) {
+        throw new BadRequestException('User not found');
+      }
+
+      // Check if a profile for this user already exists
+      const existingProfile = await this.profileRepository.findOne({ where: { id: userId } });
+      if (existingProfile) {
+        throw new BadRequestException('Profile already exists');
```
Comment (Cobidex) [inline src/modules/profile/profile.service.ts:26]: Conflict exception 409 will be more specific

File: src/modules/profile/profile.service.ts [hunk 7]
```diff
@@ -11,6 +12,36 @@ export class ProfileService {
...
+      }
+      const profile = this.profileRepository.create({ ...createProfileDto, user_id: user });
+      await this.profileRepository.save(profile);
```
Comment (Cobidex) [inline src/modules/profile/profile.service.ts:29]: store the return value in a variable for the next if-block check

File: src/modules/profile/profile.service.ts [hunk 8]
```diff
@@ -11,6 +12,36 @@ export class ProfileService {
...
+
+      const newProfile = await this.profileRepository.findOne({ where: { id: profile.id } });
```
Comment (Cobidex) [inline src/modules/profile/profile.service.ts:31]: then remove this, it doesnt seem like there is need for another db call

File: src/modules/profile/profile.service.ts [hunk 9]
```diff
@@ -11,6 +12,36 @@ export class ProfileService {
...
+
+      if (!newProfile) {
+        throw new BadRequestException('Profile creation unsuccessful');
+      }
+      return newProfile;
+    } catch (error) {
+      if (error instanceof NotFoundException || error instanceof BadRequestException) {
```
Comment (Cobidex) [inline src/modules/profile/profile.service.ts:38]: no need for try/catch block as nest wraps a try catch across all routes, the exceptions layer. any non specific error will be thrown as internal server by default

**Review by Cobidex — CHANGES_REQUESTED** [review]
Please resolve comments above
